### PR TITLE
[WIP] spark 3.0

### PR DESCRIPTION
**Pull request by Le-Zheng**
## What changes were proposed in this pull request?

Support for Spark 3.0
1. Pom: add Spark 3.0 in pom.xml, and update some related dependencies' version
2. Update Accumulator 
3. Fix match case exception, DistriOptimizerV2 serializable error, etc
## How was this patch tested?

Jenkins

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -544,7 +544,7 @@
             <id>spark_2.x</id>
             <properties>
                 <spark-version.project>2.0</spark-version.project>
-                <spark.version>2.0.0</spark.version>
+                <spark.version>2.4.0</spark.version>
```
Comment (qiuxin2012) [inline pom.xml:547]: We can use 2.4.3, the same with analytics-zoo

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/RegionProposal.scala [hunk 2]
```diff
@@ -143,9 +143,7 @@ class RegionProposal(
       val postNmsTopN = if (this.isTraining()) min(postNmsTopNTrain, bboxNumber)
       else min(postNmsTopNTest, bboxNumber)
 
-      if (output.getOrElse(b, null) == null) {
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/RegionProposal.scala:146]: why delete this? output the output this a table, we can reuse the tensors in the output.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pooler.scala [hunk 3]
```diff
@@ -121,7 +121,7 @@ class Pooler[T: ClassTag] (
       val num_rois = rois.size(1)
       totalNum += num_rois
 
-      if (out.getOrElse(i + 1, null) == null) out(i + 1) = Tensor[T]()
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pooler.scala:124]: why delete this? output the output this a table, we can reuse the tensors in the output.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/RegionProposal.scala (hunk 2 above)
Comment (Le-Zheng) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/RegionProposal.scala:146]: Here is a warning of "comparing values of types Null and Null using `==' will always yield true "

Comment (qiuxin2012) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/RegionProposal.scala:146]: Use output.contains(b) instead

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 4]
```diff
@@ -1064,6 +1064,7 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
           require(copiedModuleParamTable.get(name) != None, s"cloned module should have for $name")
           setLayerWeightAndBias(params,
             copiedModuleParamTable.get(name).get.asInstanceOf[Table], deepCopy)
+        case _=>
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala:1067]: you should throw an exception here, or mismatch will do nothing.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 5]
```diff
@@ -1125,6 +1126,7 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
         } else {
           if (matchAll) new Exception(s"module $name cannot find corresponding weight bias")
         }
+      case _=>
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala:1129]: you should throw an exception here, or mismatch will do nothing.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/caffe/CaffeLoader.scala [hunk 6]
```diff
@@ -259,6 +259,7 @@ class CaffeLoader[T: ClassTag](prototxtPath: String, modelPath: String,
     parameterTable.foreach {
       case (name: String, params: Table) =>
         copyParameter(name, params)
+      case _=>
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/caffe/CaffeLoader.scala:262]: you should throw an exception here, or mismatch will do nothing.

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/integration/torch/LSTMPeepholeSpec.scala [hunk 7]
```diff
@@ -668,6 +668,7 @@ class LSTMPeepholeSpec  extends TorchRNNSpec {
         assert(abs(v1 - v2) <= 1e-8)
         v1
       })
+    case _=>
```
Comment (qiuxin2012) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/integration/torch/LSTMPeepholeSpec.scala:671]: maybe you can delete `case ((key: Int, value: Tensor[_])) =>`, just use `asInstanceOf`

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pooler.scala [hunk 8]
```diff
@@ -121,7 +121,7 @@ class Pooler[T: ClassTag] (
       val num_rois = rois.size(1)
       totalNum += num_rois
 
-      if (out.contains(i + 1)) out(i + 1) = Tensor[T]()
+      if (!out.contains(i + 1)) out(i + 1) = Tensor[T]()
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pooler.scala:124]: why?

**Comment by Le-Zheng** [comment]
Jenkins passed Spark 2 and Spark 3 http://10.239.47.210:18080/job/BigDL-PRVN-lite/ and Integration test on yarn http://10.239.47.210:18080/view/PR/job/BigDL-PR-Integration-Examples-8/7/

**Pull request merged**